core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -4920,15 +4956,37 @@ public void rebuildDependencyGraph() {
      * @since 1.522
      */
     public Future<DependencyGraph> rebuildDependencyGraphAsync() {
-        dependencyGraphDirty.set(true);
+        synchronized (dependencyGraphLock) {
+            // Collect calls to this method to avoid unnecessary calculation of the dependency graph
+            if (scheduledFutureDependencyGraph != null) {
+                return scheduledFutureDependencyGraph;
+            }
+            // Schedule new calculation
+            return scheduledFutureDependencyGraph = scheduleCalculationOfFutureDependencyGraph(500, TimeUnit.MILLISECONDS);
+        }
+    }
+
+
+    private Future<DependencyGraph> scheduleCalculationOfFutureDependencyGraph(int delay, TimeUnit unit) {
         return Timer.get().schedule(() -> {
-            if (dependencyGraphDirty.get()) {
-                rebuildDependencyGraph();
+
+            synchronized (dependencyGraphLock) {
+                //Wait for the currently running calculation to finish
+                if (calculatingFutureDependencyGraph != null) {
```

Comment:
Because of how this works, potentially a call to `rebuildDependencyGraphAsync` will block which defeats the original purpose of the function. This needs a rethink

Reply:
Thanks yet again for your feedback. I just pushed a new commit. 

I also noticed another bug: getFutureDependencyGraph does not return the most recent dependency graph if the most recent dependency graph has been calculated synchronously. This is problematic because BuildTrigger.execute assumes that getFutureDependencyGraph will always return the most recent dependency graph. To fix this problem calculatingFutureDependencyGraph is now set to null in scheduleCalculationOfFutureDependencyGraph after the calculation has finished.